packages/perseus-score/src/util/get-scoreable-widgets.ts

```diff
@@ -0,0 +1,33 @@
+import {
+    applyDefaultsToWidgets,
+    getWidgetIdsFromContent,
+} from "@khanacademy/perseus-core";
+
+import isWidgetScoreable from "./is-widget-scoreable";
+
+import type {
+    PerseusRenderer,
+    PerseusWidgetsMap,
+} from "@khanacademy/perseus-core";
+
+/**
+ * Returns the upgraded widgets and the IDs of widgets that should be scored.
+ * Filters out widgets not referenced in content and widgets that are static or ungraded.
+ */
+export default function getScoreableWidgets(
+    perseusRenderData: PerseusRenderer,
+): {
+    upgradedWidgets: PerseusWidgetsMap;
+    scoreableWidgetIds: ReadonlyArray<string>;
+} {
+    // There seems to be a chance that PerseusRenderer.widgets might include
+    // widget data for widgets that are not in PerseusRenderer.content,
+    // so this checks that the widgets are being used before scoring them
+    const usedWidgetIds = getWidgetIdsFromContent(perseusRenderData.content);
+    // TODO: do we still need this? Shouldn't this happen during parse/migrate?
```

Comment:
I would _really_ hope so. I think the functions that take a `PerseusItem` or `PerseusRenderer` that we export to client's should accept a current version of that object. If callers suspect they may have older versions, it should be on them to  and migrate them before calling. Our types tell that story, but we still seem to "hedge our bets" in places like this.